app/src/main/java/com/lagradost/cloudstream3/utils/TvChannelUtils.kt

```diff
@@ -0,0 +1,168 @@
+import com.lagradost.cloudstream3.ui.home.HomeFragment
+import com.lagradost.cloudstream3.utils.DataStore.getKey
+import com.lagradost.cloudstream3.utils.DataStore.setKey
+import java.net.URLEncoder
+
+const val PROGRAM_ID_LIST_KEY = "persistent_program_ids"
+
+object TvChannelUtils {
+    fun Context.saveProgramId(programId: Long) {
+        val existing: List<Long> = getKey(PROGRAM_ID_LIST_KEY) ?: emptyList()
+        val updated = (existing + programId).distinct()
+        setKey(PROGRAM_ID_LIST_KEY, updated)
+    }
+    fun Context.getStoredProgramIds(): List<Long> {
+        return getKey(PROGRAM_ID_LIST_KEY) ?: emptyList()
+    }
+    fun Context.removeProgramId(programId: Long) {
+        val existing: List<Long> = getKey(PROGRAM_ID_LIST_KEY) ?: emptyList()
+        val updated = existing.filter { it != programId }
+        setKey(PROGRAM_ID_LIST_KEY, updated)
+    }
+
+
+    fun getChannelId(context: Context, channelName: String): Long? {
+        return try {
+            context.contentResolver.query(
+                TvContractCompat.Channels.CONTENT_URI,
+                arrayOf(
+                    TvContractCompat.Channels._ID,
+                    TvContractCompat.Channels.COLUMN_DISPLAY_NAME
+                ),
+                null,
+                null,
+                null
+            )?.use { cursor ->
+                while (cursor.moveToNext()) {
+                    val id = cursor.getLong(
+                        cursor.getColumnIndexOrThrow(TvContractCompat.Channels._ID)
+                    )
+                    val name = cursor.getString(
+                        cursor.getColumnIndexOrThrow(TvContractCompat.Channels.COLUMN_DISPLAY_NAME)
+                    )
+                    if (name == channelName) return id
+                }
+                null
+            }
+        } catch (e: Exception) {
+            Log.e("TvChannelUtils", "Query failed: ${e.message}", e)
+            null
+        }
+    }
+
+    /** Insert programs into a channel */
+    fun wrapWithWsrvIfNeeded(url: String?): String? {
+        if (url.isNullOrBlank()) return null
+        return if (url.contains("wsrv.nl")) {
+            url
+        } else {
+            "https://wsrv.nl/?url=${URLEncoder.encode(url, "UTF-8")}"
+        }
+    }
+
+    fun addPrograms(context: Context, channelId: Long, items: List<SearchResponse>) {
+        for (item in items) {
+            try {
+                val nameBase64 = base64Encode(item.apiName.toByteArray(Charsets.UTF_8))
+                val urlBase64 = base64Encode(item.url.toByteArray(Charsets.UTF_8))
+                val csshareUri = "$APP_STRING_SHARE:$nameBase64?$urlBase64"
+                val safePosterUrl = wrapWithWsrvIfNeeded(item.posterUrl)
+                val builder = PreviewProgram.Builder()
+                    .setChannelId(channelId)
+                    .setTitle(item.name)
+                    .setDescription(item.apiName+"Score:- ${item.score}")
```

Comment:
Same thing but localized and more robust score strings.
```suggestion
                    .apply {
                        val scoreText = item.score?.toStringNull(0.1, 10, 1)?.let {
                            " - " + txt(R.string.rating_format, it).asString(context)
                        } ?: ""
                        setDescription("${item.apiName}$scoreText")
                    }
```